src/controller/python/chip/clusters/Attribute.py

```diff
@@ -401,68 +401,73 @@ def UpdateCachedData(self):
         tlvCache = self.attributeTLVCache
         attributeCache = self.attributeCache
 
-        for endpoint in tlvCache:
+        for attributePath in changedPathSet:
+            endpoint = attributePath.EndpointId
+
             if (endpoint not in attributeCache):
                 attributeCache[endpoint] = {}
 
             endpointCache = attributeCache[endpoint]
 
-            for cluster in tlvCache[endpoint]:
-                if cluster not in _ClusterIndex:
+            cluster = attributePath.ClusterId
+
+            if cluster not in _ClusterIndex:
+                #
+                # #22599 tracks dealing with unknown clusters more
+                # gracefully so that clients can still access this data.
+                #
+                continue
+
+            clusterType = _ClusterIndex[cluster]
+
+            if clusterType not in endpointCache:
+                endpointCache[clusterType] = {}
+
+            clusterCache = endpointCache[clusterType]
+            clusterDataVersion = self.versionList.get(
+                endpoint, {}).get(cluster, None)
+
+            if (self.returnClusterObject):
+                try:
+                    # Since the TLV data is already organized by attribute tags, we can trivially convert to a cluster object representation.
+                    endpointCache[clusterType] = clusterType.FromDict(
+                        data=clusterType.descriptor.TagDictToLabelDict([], tlvCache[endpoint][cluster]))
+                    endpointCache[clusterType].SetDataVersion(
+                        clusterDataVersion)
+                except Exception as ex:
+                    decodedValue = ValueDecodeFailure(
+                        tlvCache[endpoint][cluster], ex)
+                    endpointCache[clusterType] = decodedValue
+            else:
+                clusterCache[DataVersion] = clusterDataVersion
+
+                attribute = attributePath.AttributeId
```

Comment:
suggest rename to attribute_id

Reply:
Yeah definitely helps for clarity. I guess for consistency I should also rename to `endpointId` and `clusterId`.